include/drivers/i2c.h

```diff
@@ -174,11 +175,31 @@ typedef int (*i2c_api_slave_register_t)(struct device *dev,
 typedef int (*i2c_api_slave_unregister_t)(struct device *dev,
 					  struct i2c_slave_config *cfg);
 
+typedef int (*i2c_api_recover_bus_t)(struct device *dev);
+typedef int (*i2c_api_prepare_recovery_t)(struct device *dev);
+typedef int (*i2c_api_unprepare_recovery_t)(struct device *dev);
+
+struct i2c_api_recovery_info {
```

Comment:
Do we really need all this generically?  Some drivers already know how to access the SDA and SCL pins.  Others may not even need them.  Without seeing `i2c_recover_bus_gpio()` I can't judge whether this is appropriate.

Based on my experience with [nrfcxx](https://github.com/pabigot/nrfcxx/blob/c7088a998cde7c7fec05523d36ebed1014cc05f4/src/twi.cc#L75) I think we could just get away with a single `i2c_api_recover_bus recover` function, which is responsible for doing the whole reset, or to reuse a common gpio implementation with API like this by doing any pre- and post-recovery actions around a call to an `i2c_gpio_clear_bus()` function does the bitbang loop :
```
struct gpio_pin_config_t {
  struct device *dev;
  gpio_pin_t pin;
  gpio_dts_flags_t flags;
};

int i2c_gpio_clear_bus(const struct gpio_pin_config_t *scl,
                       const struct gpio_pin_config_t *sda);
```

Reply:
@pabigot Thank you for the feedback. I have changed the PR substantially based on your comment above.

I have only implemented the new I2C API for bus recovery in the bit banging I2C library and GPIO driver for now since implementing it for other I2C controllers (at least the ones I have looked at) requires run-time pinctrl/pinmux API calls, which are being reworked.

While the API addition is limited to the GPIO bit banging I2C controller driver, this can easily be utilised from board-specific code to attempt an I2C bus recovery (the board-specific code can change the pinmux to configure SCL+SDA as GPIOs, call `i2c_recover_bus()` on the GPIO I2C controller, and reconfigure the SCL+SDA lines as I2C.

Once the pinmux/pinctrl API/DTS is in place, other I2C drivers can utilise the `i2c_bitbang.h` API implementing their own SCL+SDA setters and getters - or implement their own, controller-specific bus recovery sequence.